smartsim/_core/generation/operations.py

```diff
@@ -0,0 +1,298 @@
+    """
+    try:
+        combined_path = job_root_path
+        if dest:
+            combined_path = job_root_path / dest
+        return str(combined_path)
+    except Exception as e:
+        raise ValueError(f"Error combining paths: {e}")
+
+
+def check_src_and_dest_path(
+    src: pathlib.Path, dest: t.Union[pathlib.Path, None]
+) -> None:
+    """Validate that the provided source and destination paths are
+    of type pathlib.Path
+
+    :param src: The source path to be checked.
+    :param dest: The destination path to be checked.
+    :raises TypeError: If either src or dest is not an instance of pathlib.Path
+    """
+    if not isinstance(src, pathlib.Path):
+        raise TypeError(f"src must be of type pathlib.Path, not {type(src).__name__}")
+    if dest is not None and not isinstance(dest, pathlib.Path):
+        raise TypeError(
+            f"dest must be of type pathlib.Path or None, not {type(dest).__name__}"
+        )
+    if isinstance(dest, pathlib.Path) and not dest.is_absolute():
+        raise ValueError("Invalid destination path")
+    if isinstance(dest, pathlib.Path) and " " in str(dest):
+        raise ValueError("Path contains spaces, which are not allowed")
+
+    # TODO I want to add the check below but I do not think this works for remote jobs
+    # full_path = path.abspath(file_path)
+    #     if path.isfile(full_path):
+    #         return full_path
+    #     if path.isdir(full_path):
+    #         return full_path
+
+
+def check_run_path(run_path: pathlib.Path) -> None:
+    """Validate that the provided run path is of type pathlib.Path
+
+    :param run_path: The run path to be checked
+    :raises TypeError: If either run path is not an instance of pathlib.Path
+    :raises ValueError: If the run path is not a directory
+    """
+    if not isinstance(run_path, pathlib.Path):
+        raise TypeError(
+            f"run_path must be of type pathlib.Path, not {type(run_path).__name__}"
+        )
+    if run_path == pathlib.Path(""):
+        raise ValueError(f"Job root path '{run_path}' is not a directory.")
+
+
+class GenerationContext:
+    """Context for file system generation operations."""
+
+    def __init__(self, job_root_path: pathlib.Path):
+        check_run_path(job_root_path)
+        self.job_root_path = job_root_path
+        """The Job run path"""
+
+
+class GenerationProtocol(t.Protocol):
+    """Protocol for Generation Operations."""
+
+    def format(self, context: GenerationContext) -> Command:
+        """Return a formatted Command."""
+
+
+class CopyOperation(GenerationProtocol):
+    """Copy Operation"""
+
+    def __init__(
+        self, src: pathlib.Path, dest: t.Optional[pathlib.Path] = None
+    ) -> None:
+        """Initialize a CopyOperation object
+
+        :param src: Path to source
+        :param dest: Path to destination
+        """
+        check_src_and_dest_path(src, dest)
+        self.src = src
+        self.dest = dest
```

Comment:
Can we resolve `dest` to some relative path right away here so that we don't have to deal with the nullish possiblitly? Like could we do something like
```suggestion
        self.src = src
        self.dest = dest or src.name
```
?